### ISSUE TITLE
evidence: total evidence size should be in terms of bytes rather than quantity

### ISSUE THREAD

**Issue by cmwaters** [tendermint/tendermint#5432]
## Problem Definition

Currently, we limit the number of evidence that can make its way into a block (Evidence.MaxNum). We also set a number `MaxEvidenceBytes` (which is currently 444). This number is used to know how many bytes to allocate for transactions. The sum of these two numbers is checked to make sure that we don't exceed the max bytes of the entire block itself.

Now the newly-introduced `LightClientAttackEvidence` is considerably larger. I ran a test creating evidence in a network of 500 validators and it came out to be 87119 bytes. 

We now need to update the `MaxEvidenceBytes` to reflect this number. However if we are to use 87119 bytes as a number, then say if we propose a block with 1 DuplicateVoteEvidence we will allocate based on the LightClientAttackEvidence and thus loose (87119 - 444) bytes that could have been filled with transactions. 

Another problem that arises is say that a network has small block sizes of 200,000 Bytes (current default is 22million). This would mean that at most the user can set the Evidence.MaxNum to 2. In the case of DuplicateVoteEvidence this only takes 444 bytes and would throttle how quickly we could punish the duplicate voters.

TL;DR: having two evidence of vastly different sizes makes the max num parameter a poor mechanism

## Proposal

We should use bytes instead of quantity. This would mean that when storing evidence in the pool we save the byte size and when `PendingEvidence(maxSize int64)` is called we fill it with as much evidence we have so long as it doesn't exceed the max byte size. 

As well as returning the evidence we should also return the accumulated byte size so that the block executor can quickly work out how many txs can also fit in the block. 

Note: The reason why we might still want to have a limit on the bytes of evidence and not let it take up the entire block if need be is because we might still want proposers to be able to fit in txs as this pays gas and might be critical for the networks economic model. I do believe by default that they should be the same size i.e. by default we fill up as much of the block as we can.

This would not be any more breaking than the current change as `Evidence.MaxNum` is already new relative to 0.33.8

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by melekes** [comment]
sound good to me
